x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBytes.java

```diff
@@ -184,6 +185,37 @@ public static void copyToCacheFileAligned(
         }
     }
 
+    /**
+     * Copy all bytes from {@code input} to {@code fc}, only doing writes aligned along {@link #PAGE_SIZE}.
+     *
+     * @param fc output cache file reference
+     * @param input stream to read from
+     * @param fileChannelPos position in {@code fc} to write to
+     * @param progressUpdater callback to invoke with the number of copied bytes as they are copied
+     * @param buffer bytebuffer to use for writing
+     * @return the number of bytes copied
+     * @throws IOException on failure
+     */
+    public static int copyToCacheFileAligned(IO fc, InputStream input, int fileChannelPos, IntConsumer progressUpdater, ByteBuffer buffer)
+        throws IOException {
+        int bytesCopied = 0;
+        while (true) {
+            final int bytesRead = Streams.read(input, buffer, buffer.remaining());
+            if (bytesRead <= 0) {
+                break;
+            }
+            if (buffer.hasRemaining()) {
+                // ensure that last write is aligned on 4k boundaries (= page size)
+                final int remainder = buffer.position() % PAGE_SIZE;
+                final int adjustment = remainder == 0 ? 0 : PAGE_SIZE - remainder;
+                buffer.position(buffer.position() + adjustment);
+            }
+            bytesCopied += positionalWrite(fc, fileChannelPos + bytesCopied, buffer);
+            progressUpdater.accept(bytesCopied);
```

Comment:
Mostly for my education: With the existing version of this method, we adjust `bytesCopied` with the page alignment before invoke `progressUpdater`. I think this is beacuse `ProgressListenableActionFuture` has a few assertions that expect `bytesCopied` not exceeding its `end` value ([example](https://github.com/elastic/elasticsearch/blob/99719f2c8254e79daa182ce02ffde9c7f27baeba/x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java#L78)). 

We don't need the same adjustment here for the callback because `ProgressListenableActionFuture` should always be created with page aligned `end` value and this is guaranteed by the new `BlobCacheUtils#computeRange` method?

Reply:
> I think this is beacuse `ProgressListenableActionFuture` has a few assertions that expect `bytesCopied` not exceeding its `end` value ([example](https://github.com/elastic/elasticsearch/blob/99719f2c8254e79daa182ce02ffde9c7f27baeba/x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java#L78)).

Yes, we cannot update progress to a value that would exceed the pending range in the SparseFileTracker.

> We don't need the same adjustment here for the callback because `ProgressListenableActionFuture` should always be created with page aligned `end` value and this is guaranteed by the new `BlobCacheUtils#computeRange` method?

Yes :+1: